test/models/user_test.rb

```diff
@@ -1033,4 +1033,29 @@ def editing_self_helper
       assert_equal ['production'], User.current.visible_environments
     end
   end
+
+  context 'personal access token auth' do
+    let(:user) { FactoryGirl.create(:user) }
+    let(:token) { FactoryGirl.create(:personal_access_token, :user => user) }
+    let(:token_value) do
+      token_value = token.generate_token
+      token.save
+      token_value
+    end
+
+    test 'user can api login via personal access token' do
+      assert_nil token.last_used_at
+      assert_equal user, User.try_to_login(user.login, token_value, true)
+      assert_not_nil token.reload.last_used_at
+    end
+
+    test 'user can not ui login via personal access token' do
+      assert_nil User.try_to_login(user.login, token_value, false)
+    end
```

Comment:
1. can you add a test that expired token doesnt work when loggin in?
2. since we allow to login to the UI using the token, should we show a message if the token has expired?
3. should we allow to use tokens for the ui?
4. should we log somewhere that a user has logged via token vs regular password?

Reply:
I think, it does not make any sense to allow users to login to the UI with the token. It should be used for automated actions only.
I added the test for the expired token, will push in a second.
We already log in `User.try_to_login` when the user was authenticated with a personal access token.

Reply:
@timogoebel thanks, I missread your try to login method - thanks.